packages/runner/src/storage/interface.ts

```diff
@@ -88,3 +99,180 @@ export interface IStorageProvider {
+   */
+  status(): Result<
+    IStorageTransactionProgress,
+    IStorageTransactionError
+  >;
+
+  /**
+   * Reads a value from the the storage and captures it in the transaction
+   * invariants. Read will fail with `IStorageTransactionError` if transaction
+   * has an error state. Read will fail with `IStorageTransactionClosed` if
+   * transaction is done.
+   */
+  read(
+    address: IStorageAddress,
+  ): Result<
+    Read,
+    IStorageTransactionError | IStorageTransactionClosed
+  >;
+
+  /**
+   * Write a value into a storage at a given address & captures it in the
+   * transaction invariants. Write will fail with `IStorageTransactionError`
+   * if transaction has an error state. Write will fail with
+   * `IStorageTransactionClosed` if transaction is done.
+   */
+  write(
+    address: IStorageAddress,
+    value?: JSONValue,
+  ): Result<Write, IStorageTransactionError | IStorageTransactionClosed>;
+}
+
+/**
+ * This is transaction representation from the storage perpective. It will not
+ * be exposed outside of the storage provider intenals and is designed to allow
+ * storage provider to maintain consistency guarantees.
+ */
+export interface IStorageOpenTransaction {
+  /**
+   * This is an internal method called by a storage provider that lets
+   * transaction know about potential invariant changes. Transaction can track
+   * of all the changes internally and if any of the changes affect any of it's
+   * invariants it can transition transaction state from `Open` to failed with
+   * `IStorageConsistencyError`.
+   */
+  merge(changes: Iterable<State>): void;
+}
+
+/**
+ * Error that is produced when transaction is being updated after it was already
+ * aborted.
+ */
+export interface IStorageTransactionAborted extends Error {
+  /**
+   * Reason provided when transaction was aborted.
+   */
+  reason: unknown;
+}
+
+/**
+ * Error indicates that transaction consistency guarantees have being
+ * invalidated - some fact has changed while transaction was in progress.
+ */
+export interface IStorageConsistencyError extends Error {}
+
+export type IStorageTransactionError =
+  | IStorageTransactionAborted
+  | ConflictError
+  | TransactionError
+  | ConnectionError
+  | AuthorizationError;
+
+export interface IStorageTransactionClosed extends Error {}
+
+export type IStorageTransactionProgress = Variant<{
+  open: IStorageTransactionLog;
+  pending: IStorageTransactionLog;
+  done: IStorageTransactionLog;
+}>;
+
+export interface IStorageAddress {
+  the: The;
+  of: Entity;
+  at: string[];
+}
+
+export interface IStorageTransactionLog
+  extends Iterable<IStorageTransactionInvariant> {
+  get(address: IStorageAddress): IStorageTransactionInvariant;
+}
+
+export type IStorageTransactionInvariant = Variant<{
+  read: Read;
+  write: Write;
+}>;
+
+export interface Read {
+  readonly the: The;
+  readonly of: Entity;
+  readonly at: string[];
+  readonly is?: JSONValue;
+  readonly cause: Reference;
```

Comment:
Should `cause` be here? I was expecting cause to be managed by the transaction itself. I.e. `Read` and `Write` here don't have `cause`. And it's the responsibility of the user of this API to not use data between transactions.

Reply:
Ah never mind, I misread this earlier. This is just the return value, not the input.